### ISSUE TITLE
[BUG] Error on startup when using Oracle datasource.

### ISSUE THREAD

**Issue by dannywalk** [jobrunr/jobrunr#150]
**Are you running the latest version of JobRunr?**
Yes 

**Describe the bug**
NullPointerException On startup of my springboot app

I *think* this is because it's trying to load data from the JOBRUNR_JOBS_STATS view, and expecting the various fields to be LONG, but one of them (succedded) is coming back as NULL.  This seems to kill startup of my springboot app.

**Environment**
I'm using JobRunr version: org.jobrunr:jobrunr-spring-boot-starter:3.1.1
I'm running on JRE / JDK (e.g. OpenJDK 1.8.0_292):  openjdk 11.0.4 2019-07-16
I'm using the following StorageProvider (e.g. Oracle / MongoDB / ...): Oracle

**How to Reproduce**
Just start the app with data in the jobrunner tables.

**Expected behavior**
App starts up without error

**Additional context**

The NPE looks like:

```
2021-06-15 13:49:04.254  WARN 95227 --- [pool-2-thread-2] o.j.storage.AbstractStorageProvider      : Error notifying JobStorageChangeListeners

java.lang.NullPointerException: null
        at org.jobrunr.utils.reflection.autobox.Autoboxer.autobox(Autoboxer.java:29) ~[jobrunr-3.1.2.jar:na]
        at org.jobrunr.utils.reflection.ReflectionUtils.autobox(ReflectionUtils.java:288) ~[jobrunr-3.1.2.jar:na]
        at org.jobrunr.storage.sql.common.db.SqlResultSet.asLong(SqlResultSet.java:52) ~[jobrunr-3.1.2.jar:na]
        at org.jobrunr.storage.sql.common.DefaultSqlStorageProvider.toJobStats(DefaultSqlStorageProvider.java:259) ~[jobrunr-3.1.2.jar:na]
        at org.jobrunr.storage.sql.common.DefaultSqlStorageProvider.lambda$getJobStats$1(DefaultSqlStorageProvider.java:235) ~[jobrunr-3.1.2.jar:na]
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[na:na]
        at org.jobrunr.storage.sql.common.db.SqlSpliterator.tryAdvance(SqlSpliterator.java:44) ~[jobrunr-3.1.2.jar:na]
        at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127) ~[na:na]
        at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502) ~[na:na]
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488) ~[na:na]
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[na:na]
        at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150) ~[na:na]
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:na]
        at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543) ~[na:na]
        at org.jobrunr.storage.sql.common.DefaultSqlStorageProvider.getJobStats(DefaultSqlStorageProvider.java:236) ~[jobrunr-3.1.2.jar:na]
        at org.jobrunr.storage.AbstractStorageProvider.notifyJobStatsOnChangeListeners(AbstractStorageProvider.java:76) ~[jobrunr-3.1.2.jar:na]
        at org.jobrunr.storage.sql.common.DefaultSqlStorageProvider.save(DefaultSqlStorageProvider.java:133) ~[jobrunr-3.1.2.jar:na]
        at org.jobrunr.storage.ThreadSafeStorageProvider.save(ThreadSafeStorageProvider.java:94) ~[jobrunr-3.1.2.jar:na]
        at org.jobrunr.server.BackgroundJobServer.scheduleJob(BackgroundJobServer.java:181) ~[jobrunr-3.1.2.jar:na]
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183) ~[na:na]
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177) ~[na:na]
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654) ~[na:na]
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[na:na]
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[na:na]
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[na:na]
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[na:na]
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:na]
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497) ~[na:na]
        at org.jobrunr.server.JobZooKeeper.processRecurringJobs(JobZooKeeper.java:178) ~[jobrunr-3.1.2.jar:na]
        at org.jobrunr.server.JobZooKeeper.checkForRecurringJobs(JobZooKeeper.java:108) ~[jobrunr-3.1.2.jar:na]
        at org.jobrunr.server.JobZooKeeper.runMasterTasksIfCurrentServerIsMaster(JobZooKeeper.java:93) ~[jobrunr-3.1.2.jar:na]
        at org.jobrunr.server.JobZooKeeper.run(JobZooKeeper.java:76) ~[jobrunr-3.1.2.jar:na]
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[na:na]
        at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) ~[na:na]
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
        at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
```

I think it's because the code that reads the data from JOBRUNR_JOBS_STATS is expecting all the data to be autoboxed to long or int, but one of them (succeeded) is coming back as null from the view and throwing the NPE.

Specifically, line 259 of DefaultSqlStorageProvider.java:

```
    private JobStats toJobStats(SqlResultSet resultSet, Instant instant) {
        return new JobStats(
                instant,
                resultSet.asLong("total"),
                resultSet.asLong("awaiting"),
                resultSet.asLong("scheduled"),
                resultSet.asLong("enqueued"),
                resultSet.asLong("processing"),
                resultSet.asLong("failed"),
                resultSet.asLong("succeeded"), <<<<< 259 here >>>>>
                resultSet.asLong("allTimeSucceeded"),
                resultSet.asLong("deleted"),
                resultSet.asInt("nbrOfRecurringJobs"),
                resultSet.asInt("nbrOfBackgroundJobServers")
        );
    }
```

I think the bug is in the view - don't know why it returns null when there should be a non-null value.

**Comment by dannywalk** [comment]
Hi Ronald, thanks for the prompt response!

In our use case, we can't let JobRunr auto-create tables and views, we had to use the scripts.  (in the org I'm at, apps don't have DDL rights in oracle schemas so we had to have a DBA create the tables and whatnot via the scripts.)  

FWIW it seems to work correctly when we let it auto-create in an H2 database. 

In the meantime, we've disabled the dashboard and that seems to have fixed it for now (we can live without the dashboard for a while)

Happy to test out your changes, will let you know here how I get on.  Let me know when you've pushed a branch up.

**Comment by rdehuyss** [comment]
Ok, then I understand the problem. 

There is a new script for Oracle (# 010), please apply it and all should be well (dashboard will work again).

Will you need a different schema? If so, please built the following branch and report back. Once I know that works not only for me, I can merge it to master and release a new version. I would really appreciate that.

The branch: 
https://github.com/jobrunr/jobrunr/tree/feature/schema-support-in-sqlstorageprovider

**Comment by dannywalk** [comment]
I've checked out your branch and it built fine for me - i'll try adding the binaries to our springboot app to see if it fixes the problems tomorrow (it's late here in NZ)

Thanks again for your prompt attention to this.  

One other thing - we've noticed in our environment, that the JOBRUNR_METADATA table is empty (which might explain the null value in the view from before).  I think it should probably have at least one row in it in a running environment?

**Comment by rdehuyss** [comment]
Yes, that is correct. See script v009:

```
INSERT INTO jobrunr_metadata (id, name, owner, value, createdAt, updatedAt)
VALUES ('succeeded-jobs-counter-cluster', 'succeeded-jobs-counter', 'cluster',
        cast((select amount from jobrunr_job_counters where name = 'SUCCEEDED') as char(10)), CURRENT_TIMESTAMP,
        CURRENT_TIMESTAMP);
```

In a new installation this would be:
```
INSERT INTO jobrunr_metadata (id, name, owner, value, createdAt, updatedAt)
VALUES ('succeeded-jobs-counter-cluster', 'succeeded-jobs-counter', 'cluster', '0', CURRENT_TIMESTAMP,
        CURRENT_TIMESTAMP);
```

**Comment by rdehuyss** [comment]
Hi @dannywalk - were you able to test the branch which has schema support?

**Comment by dannywalk** [comment]
Hey Ronald, so sorry for late response.  We did manage to do some testing, but we encountered the same NPE after running the updated migration scripts in Oracle.  In the meantime, we've disabled the UI and the issue has gone for now.

**Comment by rdehuyss** [comment]
Hi @dannywalk , I triple checked and with me it works all fine on Oracle.

I also merged the support for a table prefix (e.g. like a schema, ...) and all works fine. I will soon release this version and I hope that all is then well for you.

**Comment by rdehuyss** [comment]
Resolved in [release v3.2.0](https://github.com/jobrunr/jobrunr/releases/tag/v3.2.0).
